### ISSUE TITLE
/proc - Dreaddir does not find pids 0 and 1

### ISSUE THREAD

**Issue by skaftetryne** [freemint/freemint#394]
One thing I noticed when writing jtop (https://atari.joska.no/jtop/) and ps (https://atari.joska.no/ps/) is that Dreaddir does not find the first two processes. The first entry found is ".", followed by ".." and then pid 2. But both pid 0 and 1 can be opened just like any other process. Is this intentional or a bug? I'd say the latter. I'm pretty sure that both pids used to be seen with a plain ls /proc earlier, as I can remember demonstrating to my Amiga owning friends at NVG how to kill my Falcon by deleting /proc/mint.000 via ftp back in the late 90's....

**Comment by th-otto** [comment]
Could maybe be caused by the patch where i added the "." and ".." entries. I'll have a look.

**Comment by skaftetryne** [comment]
Yes I think it is. Looks like you replace the two first files with . and ..

`	for (;;)
	{
		i = dirh->index++;

		if (i == 0)
		{
			strcpy(namebuf, ".");
			p = 0;
			break;
		}
		if (i == 1)
		{
			strcpy(namebuf, "..");
			p = 0;
			break;
		}
`

**Comment by MedourMehdi** [comment]
Just as a reminder switch case is better than if for gcc optimisation

**Comment by th-otto** [comment]
switch can't be used because the "break" has to get out of the loop.

**Comment by th-otto** [comment]
Should hopefully be fixed with https://github.com/freemint/freemint/commit/c6dc8f039e4a1bcff38e94dad9876213c7032e40

**Comment by th-otto** [comment]
Hm, yes, you are right, i did not notice that in my tests:

```
PID  PPID PRI CURPRI STATUS    SIZE    TIME    COMMANDPID  PPID PRI CURPRI STATUS    SIZE    TIME    COMMAND
000  -01   0     0   Wait      90112 01:14.31  MiNT 
001  000   0    20   Wait      16384 00:00.02  xaloader 
000  -01   0     0   Wait      90112 01:14.31  MiNT 
001  000   0    20   Wait      16384 00:00.02  xaloader 
002  000   0     0   I/O       90112 00:00.00  update 
003  000   0     0   I/O       90112 00:00.26  AESSYS 
004  000   0    20   I/O       90112 00:00.00  aesthred 
005  000   0     0   I/O       90112 00:00.03  XaSYS 
006  005   0     0   I/O       73728 00:00.00  st-guide 
007  005   0     0   I/O       90112 00:00.03  xcontrol 
008  000   0    20   Wait     327680 00:00.00  gluestik 
009  -01   0     0   TSR     9977856 00:01.39  fvdi 
010  005   0     0   I/O     1163264 00:02.22  toswin2 
011  005   0     0   I/O     1662976 00:00.32  desktop 
012  011 -20   -20   I/O      573440 00:00.14  taskbar 
013  010   0     0   Ready   3162112 00:00.44  bash 
016  013   0    21   Run       98304 00:00.01  ps 

```

>The latter can be fixed with a different workaround though

It would be easier to just undefine the `PROC_0_1_FIX` should that cause trouble now.

>There's nothing in the docs that has ever indicated to programmers that the /proc entries could be anything other than processes.

No, but /proc should be just a normal directory like others, and those are expected to have those two entries.

The reason to introduce those "." and ".." entries was, because several tools from the coreutils package are otherwise confused. And it would also be inconsistent, because if you do a stat() on /proc, the st_nlink member should reflect those 2 entries, plus 1 for any process.

**Comment by th-otto** [comment]
I've come up with attached patch now. Could you please test it? It is supposed to handle all situations (old kernels, ones with wrong fix from 2020, and new ones)

[ps.patch.txt](https://github.com/user-attachments/files/21369881/ps.patch.txt)

**Comment by th-otto** [comment]
No problem, pushed now.

BTW, is there a reason why you don't offer also the sources of jtop?

**Comment by mikrosk** [comment]
Ah, nice. I'm a bit worried about this duality, i.e. every time joska makes an update, we have to follow the suit (and I know for a fact that joska doesn't like to use gcc and cross compilers for atari development ;-)).

**Comment by skaftetryne** [comment]
> BTW, is there a reason why you don't offer also the sources of jtop?

Yes, several.

- As long as I'm developing/supporting the project in question I don't want to "lose control" by possibly having variants out there that's been modified by others.
- I don't want to be exposed as the shitty programmer I am. Almost all of my projects has started with me exploring something interesting, ending up with a sketchy prototype. Sometimes the project turns out to be useful, and as every other shitty developer out there I *don't* discard the sketchy prototype and develop something proper, but builds on that prototype until it works well enough. jtop is such a project.

Once I decide to not develop a project further I will publish the sources.

**Comment by th-otto** [comment]
Just for info, i just pushed a workaround for SLBs. SLBs dffer in  two ways from other processes:
- Fcntl(PLOADINFO) returns the name of the loading application in the filename of ploadinfo, not the name of the SLB
- the cmdlin starts at offset 0 and has no length byte

Without the patch, i get
```
$ ps
PID  PPID PRI CURPRI STATUS    SIZE    TIME    COMMAND
028  011   0     0   I/O     1425408 00:00.86  zview 
047  000   0    20   Stop      16384 00:00.00  zvgod :\atari\zview\zview\codecs\zvgod.slb
133  000   0    20   Stop      16384 00:00.00  zvcolr :\atari\zview\zviPRIO_PROCESSew\codecs\zvcolr.slb

$ ps -f
PID  PPID PRI CURPRI STATUS    SIZE    TIME    COMMAND
028  011   0     0   I/O     1425408 00:00.86  H:\atari\zview\zview\zview.app 
047  000   0    20   Stop      16384 00:00.00  H:\atari\zview\zview\zview.app :\atari\zview\zview\codecs\zvgod.slb
133  000   0    20   Stop      16384 00:00.00  H:\atari\zview\zview\zview.app :\atari\zview\zview\codecs\zvcolr.slb
```

With the patch:
```
$ ps
PID  PPID PRI CURPRI STATUS    SIZE    TIME    COMMAND
028  011   0     0   I/O     1425408 00:00.86  zview 
047  000   0    20   Stop      16384 00:00.00  zvgod 
133  000   0    20   Stop      16384 00:00.00  zvcolr 

$ ps -f
PID  PPID PRI CURPRI STATUS    SIZE    TIME    COMMAND
028  011   0     0   I/O     1425408 00:00.86  H:\atari\zview\zview\zview.app 
047  000   0    20   Stop      16384 00:00.00  H:\atari\zview\zview\codecs\zvgod.slb 
133  000   0    20   Stop      16384 00:00.00  H:\atari\zview\zview\codecs\zvcolr.slb 
```

Maybe you want to add a similar workaround in jtop.

**Comment by skaftetryne** [comment]
Yes, I will do that. I'll just have to install something that use SLB's first.

**Comment by th-otto** [comment]
Yes, latest snapshot is here: https://tho-otto.de/snapshots/zview/
